File: app/src/main/java/com/bytecoders/androidtv/iptvservice/presenter/TrackInfoCardPresenter.java
package com.bytecoders.androidtv.iptvservice.presenter;

/*
 * Copyright (c) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.view.ViewGroup;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.bytecoders.androidtv.iptvservice.R;
import com.bytecoders.androidtv.iptvservice.m3u8parser.data.Track;

import androidx.core.content.ContextCompat;
import androidx.leanback.widget.ImageCardView;
import androidx.leanback.widget.Presenter;


/*
 * A TrackInfoCardPresenter is used to generate Views and bind Objects to them on demand.
 * It contains an Image CardView
 */
public class TrackInfoCardPresenter extends Presenter {
    private int mSelectedBackgroundColor = -1;
    private int mDefaultBackgroundColor = -1;
    private Drawable mDefaultCardImage;

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent) {
        mDefaultBackgroundColor =
                ContextCompat.getColor(parent.getContext(), R.color.black);
        mSelectedBackgroundColor =
                ContextCompat.getColor(parent.getContext(), R.color.dark_grey);
        mDefaultCardImage = parent.getResources().getDrawable(R.drawable.android_48dp, null);

        ImageCardView cardView = new ImageCardView(parent.getContext()) {
            @Override
            public void setSelected(boolean selected) {
                updateCardBackgroundColor(this, selected);
                super.setSelected(selected);
            }
        };

        cardView.setFocusable(true);
        cardView.setFocusableInTouchMode(true);
        updateCardBackgroundColor(cardView, false);
        return new ViewHolder(cardView);
    }

    private void updateCardBackgroundColor(ImageCardView view, boolean selected) {
        int color = selected ? mSelectedBackgroundColor : mDefaultBackgroundColor;

        // Both background colors should be set because the view's
        // background is temporarily visible during animations.
        view.setBackgroundColor(color);
        view.findViewById(R.id.info_field).setBackgroundColor(color);
    }

    @Override
    public void onBindViewHolder(Presenter.ViewHolder viewHolder, Object item) {
        Track track = (Track) item;

        ImageCardView cardView = (ImageCardView) viewHolder.view;
        cardView.setTitleText(track.getExtInfo().getTitle());
        //cardView.setContentText("text");

        // Set card size from dimension resources.
        Resources res = cardView.getResources();
        int width = res.getDimensionPixelSize(R.dimen.card_width);
        int height = res.getDimensionPixelSize(R.dimen.card_height);
        cardView.setMainImageDimensions(width, height);
        cardView.setMainImage(mDefaultCardImage);

        if (track.getExtInfo().getTvgLogoUrl() != null) {
            Glide.with(cardView.getContext())
                    .load(track.getExtInfo().getTvgLogoUrl())
                    .apply(RequestOptions.errorOf(mDefaultCardImage))
                    .into(cardView.getMainImageView());
        }
    }

    @Override
    public void onUnbindViewHolder(Presenter.ViewHolder viewHolder) {
        ImageCardView cardView = (ImageCardView) viewHolder.view;

        // Remove references to images so that the garbage collector can free up memory.
        cardView.setBadgeImage(null);
        cardView.setMainImage(null);
    }
}

File: tvapp/src/main/java/com/bytecoders/androidtv/iptvservice/presenter/TrackInfoCardPresenter.kt
package com.bytecoders.androidtv.iptvservice.presenter

import android.graphics.drawable.Drawable
import android.view.ViewGroup
import androidx.leanback.widget.ImageCardView
import androidx.leanback.widget.Presenter
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.bytecoders.androidtv.iptvservice.R
import com.bytecoders.m3u8parser.data.Track

class TrackInfoCardPresenter : Presenter() {
    private var defaultCardImage: Drawable? = null

    override fun onCreateViewHolder(parent: ViewGroup): ViewHolder {
        defaultCardImage = parent.resources.getDrawable(R.drawable.baseline_live_tv_white_48, null)
        val cardView: ImageCardView = HighlightSelectCardView(parent.context, R.color.black, R.color.dark_grey).apply {
            isFocusable = true
            isFocusableInTouchMode = true
            updateCardBackgroundColor(false)
        }
        return ViewHolder(cardView)
    }


    override fun onBindViewHolder(viewHolder: ViewHolder, item: Any) {
        val track = item as Track
        val cardView = viewHolder.view as ImageCardView
        cardView.titleText = track.extInfo!!.title
        cardView.contentText = cardView.context.getString(
                if (track.extInfo?.tvgId.isNullOrEmpty()) R.string.no_epg else R.string.epg_avail
        )
        val res = cardView.resources
        val width = res.getDimensionPixelSize(R.dimen.tv_logo_size)
        val height = res.getDimensionPixelSize(R.dimen.tv_logo_size)
        cardView.setMainImageDimensions(width, height)
        cardView.mainImage = defaultCardImage
        track.extInfo?.tvgLogoUrl?.let {
            Glide.with(cardView.context)
                    .load(it)
                    .apply(RequestOptions.errorOf(defaultCardImage))
                    .into(cardView.mainImageView)
        }
    }

    override fun onUnbindViewHolder(viewHolder: ViewHolder) {
        (viewHolder.view as? ImageCardView)?.apply {
            // Remove references to images so that the garbage collector can free up memory.
            badgeImage = null
            mainImage = null
        }
    }
}